tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/internal/findandreplace/WidgetExtractor.java

```diff
@@ -0,0 +1,84 @@
+/*******************************************************************************
+ * Copyright (c) 2024 Vector Informatik GmbH and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Vector Informatik GmbH - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.ui.internal.findandreplace;
+
+import static org.junit.Assert.assertFalse;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.swt.widgets.Button;
+import org.eclipse.swt.widgets.Combo;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.ToolBar;
+import org.eclipse.swt.widgets.ToolItem;
+import org.eclipse.swt.widgets.Widget;
+
+import org.eclipse.ui.internal.findandreplace.overlay.HistoryTextWrapper;
+
+public final class WidgetExtractor {
+
+	private final Composite rootContainer;
+
+	private final String idDataKey;
+
+	public WidgetExtractor(String idDataKey, Composite container) {
+		this.idDataKey= idDataKey;
+		this.rootContainer= container;
+	}
+
+	public HistoryTextWrapper findHistoryTextWrapper(String id) {
+		return findWidget(rootContainer, HistoryTextWrapper.class, id);
+	}
+
+	public Combo findCombo(String id) {
+		return findWidget(rootContainer, Combo.class, id);
+	}
+
+	public Button findButton(String id) {
+		return findWidget(rootContainer, Button.class, id);
+	}
+
+	public ToolItem findToolItem(String id) {
+		return findWidget(rootContainer, ToolItem.class, id);
+	}
+
+	private <T extends Widget> T findWidget(Composite container, Class<T> type, String id) {
+		List<T> widgets= findWidgets(container, type, id);
+		assertFalse("more than one matching widget found for id '" + id + "':" + widgets, widgets.size() > 1);
+		return widgets.isEmpty() ? null : widgets.get(0);
+	}
+
+	@SuppressWarnings("unchecked")
```

Comment:
can we narrow down the unchecked warning into a smaller method?

```java
	private <T extends Widget> List<T> findWidgets(Composite container, Class<T> type, String id) {
		List<Widget> children= new ArrayList<>();
		children.addAll(List.of(container.getChildren()));
		if (container instanceof ToolBar toolbar) {
			children.addAll(List.of(toolbar.getItems()));
		}
		List<T> result= new ArrayList<>();
		for (Widget child : children) {
			if (type.isInstance(child)) {
				if (id.equals(child.getData(idDataKey))) {
					addChildIntoList(result, child);
				}
			}
			if (child instanceof Composite compositeChild) {
				result.addAll(findWidgets(compositeChild, type, id));
			}
		}
		return result;
	}

	@SuppressWarnings("unchecked")
	private <T extends Widget> void addChildIntoList(List<T> result, Widget child) {
		result.add((T) child);
	}
```

This might be overly verbose, but I believe it adds the benefit of being very explicit about how and which warning we violate - and it makes sure nothing else in the findWidgets method is being broken

Reply:
Valid point. It would also be possible to simply attach the annotation to a single line instead of the whole method. But since we have the class at hand anyway, I have replaced the line producing a warning with a cast operation of that class.